File: ascii_fields/animations/qfield.py
import math
import random

from ..core import Animation, clamp, render_field
from ..noise import fbm


class _Pair:
  """A virtual particle-antiparticle pair: born, drift apart, recombine, flash."""
  __slots__ = ("x", "y", "angle", "t0", "life", "reach")

  def __init__(self, x, y, angle, t0, life, reach):
    self.x = x
    self.y = y
    self.angle = angle
    self.t0 = t0
    self.life = life
    self.reach = reach


class QuantumFieldAnimation(Animation):
  """Vacuum fluctuations: a rippling energy field on which particle-antiparticle
  pairs continually bubble into existence, separate, then annihilate."""

  thresholds = [
    (0.10, " "), (0.20, "."), (0.32, ":"), (0.44, "-"), (0.56, "="),
    (0.68, "+"), (0.80, "*"), (0.90, "#"), (1.01, "@"),
  ]

  SPAWN_DT = 0.22

  def __init__(self):
    self._pairs = []
    self._next = 0.0
    self._last_elapsed = 0.0
    self._rng = random.Random()

  def render(self, width, height, elapsed, phase, options):
    if elapsed < self._last_elapsed:
      self._pairs = []
      self._next = 0.0
    self._last_elapsed = elapsed

    while elapsed >= self._next:
      self._pairs.append(_Pair(
        self._rng.uniform(0.05, 0.95),
        self._rng.uniform(0.05, 0.95),
        self._rng.uniform(0.0, math.tau),
        self._next,
        self._rng.uniform(0.5, 1.1),
        self._rng.uniform(0.04, 0.10),
      ))
      self._next += self.SPAWN_DT * self._rng.uniform(0.4, 1.6)
    self._pairs = [p for p in self._pairs if elapsed - p.t0 < p.life]

    aspect = width / max(1, height * 2.0)
    contrast = options.contrast
    # precompute pair endpoints for this frame
    active = []
    for p in self._pairs:
      age = (elapsed - p.t0) / p.life          # 0..1
      sep = math.sin(age * math.pi) * p.reach  # out then back in
      ca, sa = math.cos(p.angle), math.sin(p.angle)
      intensity = 0.5 + 0.5 * math.sin(age * math.pi)
      if age > 0.85:
        intensity += (age - 0.85) / 0.15       # annihilation flash
      active.append((p.x + ca * sep, p.y + sa * sep, p.x - ca * sep, p.y - sa * sep, intensity))

    grid = []
    for row in range(height):
      v = row / max(1, height - 1)
      line = []
      for col in range(width):
        u = col / max(1, width - 1)
        field = 0.14 + 0.30 * fbm(u * 6.0 + elapsed * 0.4, v * 6.0 - elapsed * 0.3, octaves=4)
        spark = 0.0
        for (ax, ay, bx, by, inten) in active:
          dxa = (u - ax) * aspect
          dya = v - ay
          dxb = (u - bx) * aspect
          dyb = v - by
          spark += inten * math.exp(-(dxa * dxa + dya * dya) / 0.0008)
          spark += inten * math.exp(-(dxb * dxb + dyb * dyb) / 0.0008)
        line.append(clamp((field + spark) * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/qfield.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use crate::noise::fbm;
use rand::{Rng, SeedableRng};
use rand_pcg::Pcg32;
use std::f64::consts::TAU;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "nebula" };
const TH: &[(f64, char)] = &[
  (0.10, ' '),
  (0.20, '.'),
  (0.32, ':'),
  (0.44, '-'),
  (0.56, '='),
  (0.68, '+'),
  (0.80, '*'),
  (0.90, '#'),
  (1.01, '@'),
];
const SPAWN_DT: f64 = 0.22;

#[derive(Clone, Copy)]
struct Pair {
  x: f64,
  y: f64,
  angle: f64,
  t0: f64,
  life: f64,
  reach: f64,
}

pub struct QField {
  pairs: Vec<Pair>,
  next: f64,
  last: f64,
  rng: Pcg32,
}
impl Default for QField {
  fn default() -> Self {
    Self { pairs: Vec::new(), next: 0.0, last: 0.0, rng: Pcg32::from_entropy() }
  }
}

impl Animation for QField {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    if ctx.elapsed < self.last {
      self.pairs.clear();
      self.next = 0.0;
    }
    self.last = ctx.elapsed;
    while ctx.elapsed >= self.next {
      self.pairs.push(Pair {
        x: self.rng.gen_range(0.05..0.95),
        y: self.rng.gen_range(0.05..0.95),
        angle: self.rng.gen_range(0.0..TAU),
        t0: self.next,
        life: self.rng.gen_range(0.5..1.1),
        reach: self.rng.gen_range(0.04..0.10),
      });
      self.next += SPAWN_DT * self.rng.gen_range(0.4..1.6);
    }
    self.pairs.retain(|p| ctx.elapsed - p.t0 < p.life);
    let ax = ctx.width as f64 / (ctx.height as f64 * 2.0).max(1.0);
    let contrast = ctx.options.contrast;
    let mut active: Vec<(f64, f64, f64, f64, f64)> = Vec::with_capacity(self.pairs.len());
    for p in &self.pairs {
      let age = (ctx.elapsed - p.t0) / p.life;
      let sep = (age * std::f64::consts::PI).sin() * p.reach;
      let (ca, sa) = (p.angle.cos(), p.angle.sin());
      let mut intensity = 0.5 + 0.5 * (age * std::f64::consts::PI).sin();
      if age > 0.85 {
        intensity += (age - 0.85) / 0.15;
      }
      active.push((p.x + ca * sep, p.y + sa * sep, p.x - ca * sep, p.y - sa * sep, intensity));
    }
    let mut grid = vec![0.0_f64; ctx.width * ctx.height];
    let dw = (ctx.width.saturating_sub(1)).max(1) as f64;
    let dh = (ctx.height.saturating_sub(1)).max(1) as f64;
    for row in 0..ctx.height {
      let v = row as f64 / dh;
      let base = row * ctx.width;
      for col in 0..ctx.width {
        let u = col as f64 / dw;
        let field = 0.14 + 0.30 * fbm(u * 6.0 + ctx.elapsed * 0.4, v * 6.0 - ctx.elapsed * 0.3, 4);
        let mut spark = 0.0_f64;
        for &(ax_, ay_, bx_, by_, inten) in &active {
          let dxa = (u - ax_) * ax;
          let dya = v - ay_;
          let dxb = (u - bx_) * ax;
          let dyb = v - by_;
          spark += inten * (-(dxa * dxa + dya * dya) / 0.0008).exp();
          spark += inten * (-(dxb * dxb + dyb * dyb) / 0.0008).exp();
        }
        grid[base + col] = clamp((field + spark) * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
